plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts

```diff
@@ -83,14 +91,15 @@ export class TaskWorker {
     return new TaskWorker({
       taskBroker: taskBroker,
       runners: { workflowRunner },
+      concurrentTasksLimit,
     });
   }
 
   start() {
     (async () => {
       for (;;) {
         const task = await this.options.taskBroker.claim();
-        await this.runOneTask(task);
+        this.taskQueue.add(() => this.runOneTask(task));
```

Comment:
We're not `await`ing anything here, which we'll need to do. We need something where we wait if the queue is full, but otherwise we submit the task and move on to the next one.

Reply:
we cannot await this `taskQueue.add` method call. Through experiementation, by `await` the `add` method would effectively make the `TaskWorker` an one-by-one blocking task execution again.

The mental model of the current code is, we wait for new tasks with `claim` (so it's not an infinite loop), then we just add them to the queue without limit nor waiting, and it's up to the queue to orchestrate the starting of the tasks with the configured `concurrentTasksLimit`.

Reply:
We don't want to be claiming tasks that we can't process yet though

Reply:
Gotcha! Thanks for the clarification. I will use the [`empty` event on PQueue](https://github.com/sindresorhus/p-queue#empty) to claim only when the queue is empty.

Reply:
... Actually we can't use `empty` because it's [only available in 7.3.0 of p-queue](https://github.com/sindresorhus/p-queue/releases/tag/v7.3.0), and we can't use ^7.0 because [it's ESM-only](https://github.com/sindresorhus/p-queue#install). I'll look into other options.

Reply:
I added waiting `onEmpty` before TaskWorkers tries to claim new Tasks, so the Worker will block and only continue if the queue is empty.